Change Portal SSO to special case of federation

diurnalist opened the pull request:
This updates Horizon such that the existing Portal SSO implementation is expressed as only a minimal change to the built-in websso mechanisms. The benefit is that it is rendered in the federation login method chooser that Horizon natively provides, and we can leverage all the existing flows as normal. This was possible because the Portal SSO solution works almost exactly the same way as Keystone federation, it just doesn't include a few steps in Keystone.

There were two main changes to make this possible:

- Horizon now accepts a `WEBSSO_DEFAULT_REDIRECT_URL`; if present, this URL will be used directly in the "default redirect" case. It need not point to Keystone (for Portal SSO, it does not!).
- Add a new endpoint that drops a special cookie, which is used to signal an opt-in into the new federation login flow. If this cookie is not set, the default redirect mechanism (included in Keystone since Stein) is used, which has the same experience as we currently have, where the user is transparently taken to Portal and back again.

**This does not require any changes to Portal to work!**

Commit:
Rewrite custom SSO as edge case of Keystone SSO

It turns out that, upon close inspection, our custom SSO flow is
remarkably similar to the "default WebSSO redirect" flow added in the
Stein release of Horizon. We just need to add support for one bit:
changing the redirect mechanism to allow an operator to configure an
absolute URL, as opposed to one created by a Keystone/IdP protocol pair.
The rest of the flow is the same: the user is redirected to a page via a
GET request (with the inbound origin appended as a param), and then the
/websso endpoint in Horizon takes a POST request from the IdP with the
Keystone token of the user, which is then used to initialize the user's
session.

This also adds support for a way to specify, even in the IdP chooser,
that you want to use the default redirect behavior--this is accomplished
by using "None" as the IdP in the WEBSSO_IDP_MAPPING, like so:

WEBSS_IDP_MAPPING = {
  'default': (None, None),
  'some_idp': ('idp', 'saml'),
}

Commit:
Add hack for Portal 'host' param

Portal should be updated to do matching based on this 'origin' param to
make things simpler from Horizon's PoV.

Commit:
Fixup typo

Commit:
Add opt-in endpoint for new WebSSO flow

Commit:
Fixup

Commit:
Don't show error if bypassing default redirect

Instead, go back to the login page, since we didn't actually use the
redirect flow.

Commit:
Loosen default redirect check

diurnalist commented on openstack_auth/utils.py:
```diff
@@ -242,10 +265,17 @@ def get_websso_url(request, auth_url, websso_auth):
+            # If the IdP is not found (or was explicitly undefined), and
+            # there is a default URL set, prefer it over WebSSO by protocol.
+            host = request.get_host()
+            url = ('%s?origin=%s&host=%s' %
```
Note: the `host` part is ONLY so Portal doesn't have to be updated. If we update Portal to accept either `host` or `origin` (preferring origin, which is more precise), then we can drop this.

xandroc replied:
The host param is a callback of sorts, used to verify the host is trusted and in conjunction with `next` lets us redirect users to a specific page in horizon. It looks like this functionality is lost, at least with portal sso and these new changes, can we still deep-link to a specific page like `/project/images`?

diurnalist replied:
Good call, I knew there was built-in `next` support in openstack_auth but I wasn't sure how it worked and didn't test it, thanks for the reminder. It looks like it's indeed not working, I'll have to see how to support it.

diurnalist replied:
Fixed in 0327103

diurnalist commented on openstack_auth/views.py:
```diff
@@ -58,12 +58,18 @@ def login(request):
-               (region, protocol, origin))
+        if utils.get_websso_default_redirect_url():
+            host = request.get_host()
+            url = ('%s?origin=%s&host=%s' % (
```
ditto

diurnalist commented on openstack_dashboard/cc_web_sso_urls.py:
```diff
@@ -1,19 +1,23 @@
+# the "built-in" URL in Horizon (/websso)
+if utils.is_websso_enabled():
+    urlpatterns += [
+        url(r"^ccwebsso/$", views.websso, name='chameleon_websso'),
```
We have to preserve this in order to keep Portal unchanged. If we want, we can update Portal to start redirecting to the "standard" websso endpoint, which I believe is `/auth/websso`

Commit:
Fix typo

xandroc commented on openstack_dashboard/cc_web_sso_urls.py:
```diff
@@ -1,19 +1,23 @@
-if utils.is_websso_enabled():
-    urlpatterns = [
-        url(r"^login/$", cc_websso_views.login, name='login'),
-        url(r"^logout/$", cc_websso_views.logout, name='logout'),
```
Without the ability to log out of a Horizon site independently of the portal, we lose much of the utility we gain from sso. A user logging out of horizon may still have work left to do on portal (and in fact may be doing so on another tab that now lost its session), also, logging out of portal entirely creates a barrier when going back to Horizon on either chi@tacc or chi@uc. Local logout is purposeful and enhances the user experience, it might be worth taking a look at restoring the feature in some capacity

diurnalist replied:
It should be possible to restore the "only logout locally" functionality with configuration; right now I have set up `WEBSSO_DEFAULT_REDIRECT_LOGOUT` to point to Portal's logout page. If I remove that configuration value, local logout works (though, it doesn't show a custom-tailored page indicating they can log out of Portal too, though I'm not sure this makes sense in the new SSO model anyways, as both Portal and Keystone(s) are themselves client of another central identity gateway.)

diurnalist replied:
I can confirm this works! Just tried out it on dev. I'll leave this "default logout" functionality disabled with a comment on how to enable it if we ever want to do that.

diurnalist replied:
Blarggg, but if you use the old SSO method, it will make it so you can't really log out, because it takes you to the login page, which automatically tries to log you in :/

diurnalist replied:
In the new federated model, it will be possible to log out of Horizon without logging out of the central IdP, which I think is what you're talking about. I can hack in the logout page again, so that users on the old system will not automatically have their portal session logged out if they log out of Horizon.

xandroc replied:
Agree, restoring the logout pages and pointing there when using the old regular portal sso should work. With a couple of minor changes, we may be able to keep the stand-alone logout  page when using keycloak to let users know that they may still be logged in to keycloak or other auth providers they may have used through keycloak like, `https://identity.tacc.cloud/`

xandroc replied:
Although, since Keycloak logout is working as-is, the decision of whether to integrate the logout page in federated auth could certainly be pushed out to a later release

diurnalist replied:
Fixed in 807b319 - I copied much of what you had for the original logout page (will give you a co-authored-by on these) and added a new setting to allow operators to say "yes, I support external logout, but users must confirm they want that." If that's set, then instead of automatically logging them out from the IdP (Portal in this case), we show the logout page with a link to logout from the IdP if they desire.

Commit:
Support redirect 'next' param in WebSSO

If a user logs in via WebSSO, the 'next' parameter is not respected, and
so the user ultimately must re-navigate to their target destination once
they complete their login. This fixes that by including the 'next'
parameter on the Keystone WebSSO start request. Keystone will ignore
this flag. However, we can then read it from the referer when Horizon is
handling its final part of the login flow.

Commit:
Add page to display if no default logout URL

If there is a default redirect URL, but no default logout URL, the user
can effectively not logout, as they will be brought back to the login
page, which will automatically log them in again. This adds a logout
page with some overriddable messaging, and a link to do a "deep" logout

Commit:
Move bypass logic into getter

It is too easy to forget to include the bypass in a conditional.

xandroc reviewed (approved):
Looks Good!

Commit:
Prevent margin collapse on splash page

Currently the #splash body will have its margin collapsed due to the
margin-top of the .login container. This means that the background image
of #splash, if set, will be offset to the top of the .login div. To fix,
change the top padding to "padding" instead of "margin".

Commit:
Support passing redirect on websso as query param

Instead of extracting from the referer, sometimes the triggering service
may choose to include this directly.
